### PR TITLE
www-misc/buku: verbump to 4.2.2

### PR DESCRIPTION
**Pull request by stkw0**

### CODE REVIEW

**Comment by juippis** [comment]
Usually version bump comes before removal of old ebuilds, but I'm not sure if it matters in this case (keep that in mind, though, and prepare to fix it).

File: www-misc/buku/buku-4.0.ebuild [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Copyright 1999-2018 Gentoo Foundation
+# Copyright 1999-2018 Gentoo Authors
 # Distributed under the terms of the GNU General Public License v2
 
 EAPI=6
```
Comment (juippis) [inline www-misc/buku/buku-4.0.ebuild]: Looks to me that 7 can be used in here.

File: www-misc/buku/buku-4.0.ebuild [hunk 2]
```diff
@@ -9,26 +9,48 @@ inherit bash-completion-r1 distutils-r1
...
-KEYWORDS="~amd64 ~x86"
 IUSE="test"
 
 RDEPEND="
 	dev-python/beautifulsoup:4[${PYTHON_USEDEP}]
+	dev-python/certifi[${PYTHON_USEDEP}]
 	dev-python/cryptography[${PYTHON_USEDEP}]
+	dev-python/html5lib[${PYTHON_USEDEP}]
 	>=dev-python/urllib3-1.13.1[${PYTHON_USEDEP}]
+
```
Comment (juippis) [inline www-misc/buku/buku-4.0.ebuild]: Unnecessary extra line, also please sort these deps in alphabetical order for better readability.

Comment (stkw0) [reply www-misc/buku/buku-4.0.ebuild]: buku have two programs buku and bukuserver. The first dependencies are of buku and the second set are of bukuserver, this way it's easier to keep track of bukuserver dependencies.

Comment (juippis) [reply www-misc/buku/buku-4.0.ebuild]: Okay, I see. That's kind of smart, even though if both are shipped unconditionally, does it really matter in separating them? But wait for a dev to give their input. I've never seen RDEPEND list like this before, but it doesn't mean they don't exist! I'd personally probably try something like 
```
CLIENT_DEPEND="dev-python/beautifulsoup:4[${PYTHON_USEDEP}]
	dev-python/certifi[${PYTHON_USEDEP}]
	dev-python/cryptography[${PYTHON_USEDEP}]
	dev-python/html5lib[${PYTHON_USEDEP}]
	>=dev-python/urllib3-1.13.1[${PYTHON_USEDEP}]"
SERVER_DEPEND="dev-python/arrow[${PYTHON_USEDEP}]
	dev-python/click[${PYTHON_USEDEP}]
	dev-python/flask[${PYTHON_USEDEP}]
	dev-python/flask-admin[${PYTHON_USEDEP}]
	dev-python/flask-api[${PYTHON_USEDEP}]
	dev-python/flask-bootstrap[${PYTHON_USEDEP}]
	dev-python/flask-paginate[${PYTHON_USEDEP}]
	dev-python/flask-wtf[${PYTHON_USEDEP}]
	dev-python/requests[${PYTHON_USEDEP}]
	dev-python/werkzeug[${PYTHON_USEDEP}]"
RDEPEND="${CLIENT_DEPEND}
	${SERVER_DEPEND}"
DEPEND="${CLIENT_DEPEND}
	test? (
		dev-python/attrs[${PYTHON_USEDEP}]
		dev-python/flake8[${PYTHON_USEDEP}]
		dev-python/hypothesis[${PYTHON_USEDEP}]
		dev-python/py[${PYTHON_USEDEP}]
		dev-python/pytest[${PYTHON_USEDEP}]
		dev-python/pyyaml[${PYTHON_USEDEP}]
		dev-python/vcrpy[${PYTHON_USEDEP}]
	)"
```

File: www-misc/buku/buku-4.0.ebuild [hunk 3]
```diff
@@ -0,0 +1,75 @@
+# Copyright 1999-2018 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+
+PYTHON_COMPAT=( python3_{4,5,6} )
```
Comment (mgorny) [inline www-misc/buku/buku-4.0.ebuild]: Having python3.4 forces downgrade of beautifulsoup; remove it.

File: www-misc/buku/buku-4.0.ebuild [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	dev-python/flask-paginate[${PYTHON_USEDEP}]
+	dev-python/flask-wtf[${PYTHON_USEDEP}]
+	dev-python/requests[${PYTHON_USEDEP}]
+	dev-python/werkzeug[${PYTHON_USEDEP}]
+"
+
+DEPEND="${RDEPEND}
+	test? (
+		dev-python/attrs[${PYTHON_USEDEP}]
+		dev-python/flake8[${PYTHON_USEDEP}]
```
Comment (mgorny) [inline www-misc/buku/buku-4.0.ebuild]: Can't tests work without flake8? It's generally useless for our testing purposes.

File: www-misc/buku/buku-4.0.ebuild [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+if [[ ${PV} == *9999 ]]; then
+	inherit git-r3
+	EGIT_REPO_URI="https://github.com/jarun/Buku"
+else
+	# Tests are not working with the pypi version (https://github.com/jarun/Buku/issues/325)
+	SRC_URI="https://github.com/jarun/Buku/archive/v${PV}.tar.gz -> ${P}.tar.gz"
+	# SRC_URI="mirror://pypi/${PN:0:1}/${PN}/${P}.tar.gz -> ${P}.tar.gz"
+
+	# Github folder name starts with capital B
+	S="${WORKDIR}/Buku-${PV}"
```
[the remark this answers is not in the document]
Comment (stkw0) [reply www-misc/buku/buku-4.0.ebuild]: Thanks for the suggestion.

In my chase the tests don't have any failures. I will try to run the tests inside a chroot/docker.

[the remark this answers is not in the document]
Comment (mgorny) [reply www-misc/buku/buku-4.0.ebuild]: FEATURES=network-sandbox is your friend.

[the remark this answers is not in the document]
Comment (stkw0) [reply www-misc/buku/buku-4.0.ebuild]: thanks, tests that were not working with network-sandbox are now ignored.

I also added a USE flag for the server thing. I didn't know how to do it properly, so at the end I played around with the setup.py, but seems rather ugly

File: www-misc/buku/buku-4.1.ebuild [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+EAPI=7
+
+PYTHON_COMPAT=( python3_{5,6} )
+
+inherit bash-completion-r1 distutils-r1
+
+DESCRIPTION="Powerful command-line bookmark manager"
+HOMEPAGE="https://github.com/jarun/Buku"
+
```
Comment (mgorny) [inline www-misc/buku/buku-4.1.ebuild]: This empty line seems unnecessary.

File: www-misc/buku/buku-4.1.ebuild [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+SRC_URI="https://github.com/jarun/Buku/files/1319933/bookmarks.zip"
```
Comment (mgorny) [inline www-misc/buku/buku-4.1.ebuild]: This is a very bad and collision-prone filename. Please rename it to include 'buku' prefix, and some version indicator.

File: www-misc/buku/buku-4.1.ebuild [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+if [[ ${PV} == *9999 ]]; then
+	inherit git-r3
+	EGIT_REPO_URI="https://github.com/jarun/Buku"
+else
+	# Tests are not working with the pypi version (https://github.com/jarun/Buku/issues/325)
```
Comment (mgorny) [inline www-misc/buku/buku-4.1.ebuild]: This doesn't seem relevant to SRC_URI. Find a more meaningful place for it?

File: www-misc/buku/buku-4.1.ebuild [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+	SRC_URI="${SRC_URI} https://github.com/jarun/Buku/archive/v${PV}.tar.gz -> ${P}.tar.gz"
```
Comment (mgorny) [inline www-misc/buku/buku-4.1.ebuild]: `${P^}` to improve consistency with S.

File: www-misc/buku/buku-4.1.ebuild [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+	# SRC_URI="mirror://pypi/${PN:0:1}/${PN}/${P}.tar.gz -> ${P}.tar.gz"
```
Comment (mgorny) [inline www-misc/buku/buku-4.1.ebuild]: Remove this.

File: www-misc/buku/buku-4.1.ebuild [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+		dev-python/pyyaml[${PYTHON_USEDEP}]
+		dev-python/vcrpy[${PYTHON_USEDEP}]
+	)
+"
+
+python_prepare_all() {
+	distutils-r1_python_prepare_all
+
+	if ! use server; then
+		sed -i 's/, '\''bukuserver=bukuserver.server:cli'\''\]/\]/g' setup.py || die
```
Comment (mgorny) [inline www-misc/buku/buku-4.1.ebuild]: Please use at least `""` to avoid escape hell. Also, any chance to make those seds less fragile?

File: www-misc/buku/buku-4.1.ebuild (hunk 7 above)
Comment (mgorny) [reply www-misc/buku/buku-4.1.ebuild]: Also, if you need it only for tests, `test? ( ... )`.

File: www-misc/buku/buku-4.1.ebuild [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+	# Github folder name starts with capital B
+	S="${WORKDIR}/${P^}"
+
+	KEYWORDS="~amd64 ~x86"
+fi
+
+LICENSE="GPL-3"
+SLOT="0"
+IUSE="test server"
+
```
Comment (mgorny) [inline www-misc/buku/buku-4.1.ebuild]: RESTRICT="!test? ( test )"

File: www-misc/buku/buku-4.1.ebuild [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+	distutils-r1_python_prepare_all
+
+	if ! use server; then
+		sed -i 's/, '\''bukuserver=bukuserver.server:cli'\''\]/\]/g' setup.py || die
+		sed -i 's/de=\['\''tests/de=\['\''bukuserver'\'', '\''tests/g' setup.py || die
+	fi
+
+	if use test; then
+		unpack bookmarks.zip || die
+		mv Bookmarks places.sqlite "${S}/tests/test_bukuDb" || die
```
Comment (mgorny) [inline www-misc/buku/buku-4.1.ebuild]: You can `cd` there before unpacking.

File: www-misc/buku/buku-4.1.ebuild [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+	newbashcomp auto-completion/bash/buku-completion.bash "${PN}"
+
+	insinto /usr/share/zsh/site-functions
+	newins auto-completion/zsh/_buku _buku
+
+	doman buku.1
+}
+
+python_test() {
+	local skipped=("_rec" "refreshdb" "tnyfy_url" "with_url" "test_browse" \
```
Comment (mgorny) [inline www-misc/buku/buku-4.1.ebuild]: No need for backslash inside array.

File: www-misc/buku/buku-4.1.ebuild [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+                   "search" "_tag" "_compactdb" "cleardb" )
+
+	local args="not ${skipped[0]}"
+	for i in ${skipped[@]:1}; do
```
Comment (mgorny) [inline www-misc/buku/buku-4.1.ebuild]: "" around variable reference.

File: www-misc/buku/buku-4.1.ebuild [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+		args+=" and not $i"
+	done
+
+	py.test -vv -k "$args" tests/test_*.py || die
```
Comment (mgorny) [inline www-misc/buku/buku-4.1.ebuild]: `${args}` for consistency. Also it's `pytest` nowadays.

File: www-misc/buku/buku-4.1.ebuild (hunk 11 above)
Comment (stkw0) [reply www-misc/buku/buku-4.1.ebuild]: I replaced it with a patch. I think this way it is more obvious what is being changed and easier to maintain.

File: www-misc/buku/buku-4.1.ebuild [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+
+EAPI=7
+
+PYTHON_COMPAT=( python3_{5,6} )
+
+inherit bash-completion-r1 distutils-r1
+
+DESCRIPTION="Powerful command-line bookmark manager"
+HOMEPAGE="https://github.com/jarun/Buku"
+SRC_URI="test? ( https://github.com/jarun/Buku/files/1319933/bookmarks.zip -> ${P}-bookmarks.zip )"
```
Comment (mgorny) [inline www-misc/buku/buku-4.1.ebuild]: Don't bind it to specific package version when the same file is shared by multiple versions. This creates absurd `buku-9999-bookmarks.zip`.

File: www-misc/buku/buku-4.1.ebuild [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+
+	insinto /usr/share/zsh/site-functions
+	newins auto-completion/zsh/_buku _buku
+
+	doman buku.1
+}
+
+python_test() {
+	local skipped=("_rec" "refreshdb" "tnyfy_url" "with_url" "test_browse"
+"search" "_tag" "_compactdb" "cleardb" )
```
Comment (mgorny) [inline www-misc/buku/buku-4.1.ebuild]: Indent, please.

**Comment by mgorny** [comment]
Ping. Also, it seems that this needs rebase.

File: www-misc/buku/buku-4.2.2.ebuild [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+		dev-python/pyyaml[${PYTHON_USEDEP}]
+		dev-python/vcrpy[${PYTHON_USEDEP}]
+	)
+"
+
+python_prepare_all() {
+	distutils-r1_python_prepare_all
+
+	if ! use server; then
+		eapply "${FILESDIR}/buku-4.1-disable-server.patch" || die
```
Comment (mgorny) [inline www-misc/buku/buku-4.2.2.ebuild]: Conditional patching is a very bad idea. This patch is going to fail anytime upstream touches surrounding code, and if you build with USE=server you won't even notice. Also, `eapply` dies on its own.

My suggestion would be to prepare a proper configurable option for upstream (e.g. via setup.cfg), and submit it there.

File: www-misc/buku/buku-4.2.2.ebuild [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+	doman buku.1
+}
+
+python_test() {
+	local skipped=("_rec" "refreshdb" "tnyfy_url" "with_url" "test_browse"
+					"search" "_tag" "_compactdb" "cleardb" "_database" )
+
+	local args="not ${skipped[0]}"
+	for i in "${skipped[@]:1}"; do
+		args+=" and not $i"
```
Comment (mgorny) [inline www-misc/buku/buku-4.2.2.ebuild]: `${i}`

File: www-misc/buku/buku-4.2.2.ebuild (hunk 19 above)
Comment (stkw0) [reply www-misc/buku/buku-4.2.2.ebuild]: Did you mean something like this? I never used setup.cfg before. I will open a PR upstream with this patc

```
diff --git a/setup.cfg b/setup.cfg
new file mode 100644
index 0000000..9e16a70
--- /dev/null
+++ b/setup.cfg
@@ -0,0 +1,9 @@
+[options.extras_require]
+tests = tests_require_server_require
+server = server_require
+packaging = twine>=1.11.0
+
+[options.entry_points]
+console_scripts =
+   bukui = buku:main
+   bukuserver = bukuserver.server:cli
diff --git a/setup.py b/setup.py
index 7bd7cbd..e79154e 100644
--- a/setup.py
+++ b/setup.py
@@ -79,14 +79,6 @@ setup(
     ],
     packages=find_packages(exclude=['tests']),
     include_package_data=True,
-    entry_points={
-        'console_scripts': ['buku=buku:main', 'bukuserver=bukuserver.server:cli']
-    },
-    extras_require={
-        'tests': tests_require + server_require,
-        'server': server_require,
-        'packaging': ['twine>=1.11.0']
-    },
     test_suite='tests',
     tests_require=tests_require,
     keywords='cli bookmarks tag utility',
```

Comment (stkw0) [reply www-misc/buku/buku-4.2.2.ebuild]: I will try to fix this upstream for the next release. For now it's okay to have a patch?

**Comment by juippis** [comment]
ee57fb7ad0873b462fe7ecd7d8040211763dac23
